nat/traversal/nat_proxy.go

```diff
@@ -21,22 +21,150 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"sync"
 
 	log "github.com/cihub/seelog"
+	"github.com/mysteriumnetwork/node/core/ip"
 	"github.com/mysteriumnetwork/node/services"
 )
 
 const logPrefix = "[NATProxy] "
+const bufferLen = 30000
 
 // NATProxy provides traffic proxying functionality for registered services
 type NATProxy struct {
-	servicePorts map[services.ServiceType]int
+	servicePorts  map[services.ServiceType]int
+	addrLast      *net.UDPAddr
+	ipResolver    ip.Resolver
+	socketProtect func(socket int) bool
+	stop          chan struct{}
+	once          sync.Once
+}
+
+// consumerHandOff launches listener on pinger port and wait for openvpn connect
+// Read from listener socket and write to remoteConn
+// Read from remoteConn and write to listener socket
+func (np *NATProxy) consumerHandOff(consumerPort int, remoteConn *net.UDPConn) {
+	log.Info(logPrefix, "Inside consumer NAT proxy")
+
+	// reinit a stop channel, since it might be closed by previous connect
+	np.stop = make(chan struct{})
+
+	// outIP, err := np.ipResolver.GetOutboundIP()
+	laddr, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("%s:%d", "127.0.0.1", consumerPort+1))
+	if err != nil {
+		log.Error(logPrefix, "failed to get local address for consumer NAT proxy: ", err)
+		return
+	}
+
+	remoteConn.SetReadBuffer(bufferLen)
+	remoteConn.SetWriteBuffer(bufferLen)
+
+	fd, err := remoteConn.File()
+	if err != nil {
+		log.Error(logPrefix, "failed to fetch fd from: ", remoteConn)
+	}
+	defer fd.Close()
+
+	log.Info(logPrefix, "protecting socket: ", int(fd.Fd()))
+
+	np.socketProtect(int(fd.Fd()))
+
+	for {
+		log.Info(logPrefix, "waiting connect from openvpn3 client process")
+		// If for some reason consumer disconnects, new connection will be from different port
+		proxyConn, err := net.ListenUDP("udp4", laddr)
+		if err != nil {
+			log.Errorf("%sfailed to listen for consumer proxy on: %v, %v", logPrefix, laddr, err)
+			return
+		}
+
+		select {
+		case <-np.stop:
+			log.Info(logPrefix, "Stopping NATProxy handOff loop")
+			proxyConn.Close()
+			remoteConn.Close()
+			return
+		default:
```

Comment:
We can make nesting level a bit smaller with empty `default` block since the first case ends with return:
```suggestion
		default:
	}
```